index.ts

```diff
@@ -1,213 +1,170 @@
-        env,
-        auth_event,
-      );
-      if (!has_put_permission) {
-        return new Response("Unauthorized.", { status: 401 }); // 401 Unauthorized
-      }
-
-      /*
-            // This may be a faster solution than below but I'm not sure if it works with deno or not.
-			// Source: https://developers.cloudflare.com/workers/runtime-apis/web-crypto/#usage
-			let file_sha256_hex;
-			try {
-				// Fetch from origin
-				const res = await fetch(request);
-				if (!res.body) {
-					return new Response('body missing', { status: 400 }); // 400 Bad Request
-				}
-				// Create a SHA-256 digest stream and pipe the body into it
-				const digestStream = new crypto.DigestStream("SHA-256");
-				res.body.pipeTo(digestStream);
-				// Get the final result
-				const digest = await digestStream.digest;
-				// Turn it into a hex string
-				file_sha256_hex = [...new Uint8Array(digest)]
-					.map(b => b.toString(16).padStart(2, '0'))
-					.join('');
-			} catch (error) {
-                console.error(error);
-				return new Response('error generating sha256 hash', { status: 500 }); // 500 Internal Server Error
-			}
-            */
-      let file_sha256_hex;
+  
       try {
-        const res = await fetch(request);
-        if (!res.body) {
-          return new Response("body missing", { status: 400 }); // 400 Bad Request
+        if(await env.BANBOORU_BUCKET.head(params.hash)) {
+          return new Response("Conflict", { status: 409 })
         }
-        const bodyArrayBuffer = await res.arrayBuffer();
-        const bodyUint8Array = new Uint8Array(bodyArrayBuffer);
-        const sha256Uint8Array = sha256(bodyUint8Array);
-        file_sha256_hex = bytesToHex(sha256Uint8Array);
-      } catch (error) {
-        console.error(error);
-        return new Response("error generating sha256 hash", { status: 500 }); // 500 Internal Server Error
-      }
-
-      if (file_sha256_hex !== path_file_sha256_hex) {
-        return new Response(
-          "mismatch between generated sha-256 hash and path /file/<SHA-256>",
-          { status: 400 },
-        ); // 400 Bad Request
-      }
-
-      try {
-        await env.BANBOORU_BUCKET.put(file_sha256_hex, request.body);
+        const object = await env.BANBOORU_BUCKET.put(params.hash, request.body, {
+          // See https://developers.cloudflare.com/r2/api/workers/workers-api-reference/#r2putoptions
+          // R2 will check the received data SHA-256 hash to confirm received object integrity and will fail if it does not.
+          sha256: hexToBytes(params.hash).buffer
+        });
+        if (!object) {
+          if(payloadTag && payloadTag[1]) {
+            return new Response(`Unauthorized: Invalid nostr event, payload signature invalid`, { status: 401 });
+          } else {
+            return new Response("Bad Request", { status: 400 })
+          }
+        }
+  
+        const metadataEvent: Event = getBlankEvent(1063)
```

Comment:
It would be nice to add a comment here that kind 1063 is a NIP-94 File Metadata event